packages/eslint-plugin-office-addins/src/rules/load-object-before-read.ts

```diff
@@ -0,0 +1,146 @@
+        return node.parent.property.name;
+      }
+      return undefined;
+    }
+
+    function isLoadFunction(node: TSESTree.Node): boolean {
+      return (node.parent?.type === TSESTree.AST_NODE_TYPES.MemberExpression
+        && node.parent.property.type === TSESTree.AST_NODE_TYPES.Identifier
+        && node.parent.property.name === "load");
+    }
+
+    function callsGetAPIFunction(node: TSESTree.Identifier): boolean {
+      return (node.name.startsWith("get"));
+    }
+
+    function isVariableDeclaration(node: TSESTree.Node): boolean {
+      if(node.parent?.type === TSESTree.AST_NODE_TYPES.VariableDeclarator) {
+          return true;
+      }
+      return false;
+    }
+
+    function isGetVariableDeclaration(node: TSESTree.Node): boolean {
+      if(node.parent?.type === TSESTree.AST_NODE_TYPES.VariableDeclarator
+        && node.parent.init?.type === TSESTree.AST_NODE_TYPES.CallExpression
+        && node.parent.init.callee.type === TSESTree.AST_NODE_TYPES.MemberExpression
+        && node.parent.init.callee.property.type === TSESTree.AST_NODE_TYPES.Identifier) {
+          if(callsGetAPIFunction(node.parent.init.callee.property)) {
+            return true;
+          }
+      }
+      return false;
+    }
+
+    function isAssignmentExpression(node: TSESTree.Node): boolean {
+      if(node.parent?.type === TSESTree.AST_NODE_TYPES.AssignmentExpression) {
+          return true;
+      }
+      return false;
+    }
+
+    function isGetAssignmentExpression(node: TSESTree.Node): boolean {
+      if(node.parent?.type === TSESTree.AST_NODE_TYPES.AssignmentExpression
+        && node.parent.right.type === TSESTree.AST_NODE_TYPES.CallExpression
+        && node.parent.right.callee.type === TSESTree.AST_NODE_TYPES.MemberExpression
+        && node.parent.right.callee.property.type === TSESTree.AST_NODE_TYPES.Identifier) {
+          if(callsGetAPIFunction(node.parent.right.callee.property)) {
+            return true;
+          }
+      }
+      return false;
+    }
+
+    function getLoadedPropertyName(node: TSESTree.Node): string {
+      if(node.parent?.parent?.type === TSESTree.AST_NODE_TYPES.CallExpression
+        && node.parent.parent.arguments[0].type === TSESTree.AST_NODE_TYPES.Literal) {
+          return node.parent.parent.arguments[0].value as string;
+        }
+      return "error in getLoadedPropertyName";
+    }
+
+    function findLoadBeforeRead(scope: Scope) {
+      scope.variables.forEach((variable: Variable) => {
+        let loadLocation: Map <string, number> = new Map<string, number>();
+        let getFound: boolean = false;
+        variable.references.forEach((reference: Reference) => {
+          const node: TSESTree.Node = reference.identifier;
+            if (isVariableDeclaration(node)
+              || isAssignmentExpression(node)) {
+              if (isGetVariableDeclaration(node)
+                || isGetAssignmentExpression(node)) {
```

Comment:
Looks like you are duplicating work here.  You check for isVariableDeclaration and then you check for isGetVariableDeclaration.  You should be able to simplify those such that there is a single check for the get method that is used in both cases and you don't have to recheck the same things.

Reply:
Yeah,  did notice that. It is just one small overlap on `node. parent?.type` check.
The problem that I was having and the reason I chose to make the code like this is that Typescript keeps complaining about accessing a node type without checking for the type first. So I needed to make the condition on both of the functions.

The choice was: or leave it like that with function calls and this small code duplication (one check).
Or not use functions and put all the if checks here

Reply:
In the case of isNullObjectNode in test-for-null-using-isnullObject rule I ended up putting them in a complex if clause and had one shared method (taking advantage of a shared node type to examine that is more generic) for looking at the expression to determine if it was what I was looking for.  I could simplify a bit more by putting each half of the main if clause in a function, but I still wouldn't be testing the type twice.

Even if you didn't want to change much here you could simply remove the outer if clause and it would be the same.

Reply:
Sure, I managed to change the code to not have duplicate checks (and also tackled the parent issue within the same code). Now, there is no duplicate checks on this